### PR TITLE
getrawtransaction should take a bool for verbose

### PR DESCRIPTION
**Pull request by jnewbery**
As suggested by @gmaxwell in https://github.com/bitcoin/bitcoin/issues/9024 .

getrawtransaction now takes a bool for verbose. It will still take an int for verbose for back-compatibility, but issue a warning to the user.

### CODE REVIEW

**Comment by laanwj** [comment]
utACK, but please remove the deprecation warning. We may to repurpose the ints as 'verbosity level' in the future. See https://github.com/bitcoin/bitcoin/pull/8704#issuecomment-256321542 . There's zero overhead in simply accepting both without warnings or complaints.

**Comment by luke-jr** [comment]
Seems if we're going to do a verbosity level for getblock, we should just make that the "preferred" form here too?

**Comment by jnewbery** [comment]
> Seems if we're going to do a verbosity level for getblock, we should just make that the "preferred" form here too?

NACK. There are several other RPCs which take a bool for a more verbose output: getrawmempool, getmempoolancestors, getmempooldescendants, getblockheader, getblock, gettxout (where the option is includemempool). As far as I can tell, getrawtransaction is the *only* call which currently takes a num. It's less disruptive to change one RPC to use a bool than change all the others to use nums.

**Comment by jnewbery** [comment]
@laanwj - warning removed and commits squashed.

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -135,12 +135,12 @@ UniValue getrawtransaction(const JSONRPCRequest& request)
...
-            "If verbose is non-zero, returns an Object with information about 'txid'.\n"
+            "\nIf verbose is 'true', returns an Object with information about 'txid'.\n"
+            "If verbose is 'false' or omitted, returns a string that is serialized, hex-encoded data for 'txid'.\n"
 
             "\nArguments:\n"
             "1. \"txid\"      (string, required) The transaction id\n"
-            "2. verbose       (numeric, optional, default=0) If 0, return a string, other return a json object\n"
+            "2. verbose       (bool, optional, default=false) If true, return a string, other return a json object\n"
 
             "\nResult (if verbose is not set or set to 0):\n"
```
Comment (jonasschnelli) [inline src/rpc/rawtransaction.cpp]: maybe `s/or set to 0/or set to false`?

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -192,17 +192,20 @@ UniValue getrawtransaction(const JSONRPCRequest& request)
...
 
     uint256 hash = ParseHashV(request.params[0], "parameter 1");
 
+    // Accept either a bool (true) or a num (>=1) to indicate verbose output.
     bool fVerbose = false;
-    if (request.params.size() > 1)
-        fVerbose = (request.params[1].get_int() != 0);
+    if ((request.params.size() > 1) &&
+        ((request.params[1].isNum() && request.params[1].get_int() != 0) ||
+         (request.params[1].isBool() && request.params[1].get_bool())))
```
Comment (jonasschnelli) [inline src/rpc/rawtransaction.cpp]: maybe simplify with `s/(request.params[1].isBool() && request.params[1].get_bool())/request.params[1].isTrue()/`

**Comment by jonasschnelli** [comment]
utACK 837c06fe22a08d49f480ae629767c1ed8b7b9246, maybe consider nits/simplifications.

**Comment by jnewbery** [comment]
Thanks @jonasschnelli . Nits fixed and squashed.

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -192,17 +192,20 @@ UniValue getrawtransaction(const JSONRPCRequest& request)
...
 
     uint256 hash = ParseHashV(request.params[0], "parameter 1");
 
+    // Accept either a bool (true) or a num (>=1) to indicate verbose output.
     bool fVerbose = false;
-    if (request.params.size() > 1)
-        fVerbose = (request.params[1].get_int() != 0);
+    if ((request.params.size() > 1) &&
+        ((request.params[1].isNum() && request.params[1].get_int() != 0) ||
+         (request.params[1].isTrue())))
```
Comment (laanwj) [inline src/rpc/rawtransaction.cpp]: Note that isTrue doesn't *fail* if the value is not a bool, so this also accepts non-bool and non-int values. E.g.:
```
src/bitcoin-cli -regtest getrawtransaction 0000000000000000000000000000000000000000000000000000000000000000 '"b"'
error code: -5
error message:
No information available about transaction
```
Same for `{}` `[]` etc. Should probably raise an error in that case.

**Comment by laanwj** [comment]
@jnewbery this just needs my above comment fixed and it can be merged.

**Comment by jnewbery** [comment]
Thanks @laanwj . I've made the suggested change. I want to write some additional test cases to cover the new code and then I'll push the new commits.

**Comment by jnewbery** [comment]
@laanwj - as requested, `getrawtransaction` now throws an error whenever an argument is passed in which isn't either a bool or an int. I've also added test cases to rawtransaction.py to cover cover the RPC.

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/9025/commits/ce2bb23aa5be10232726d063bfae6063a14720b9

**Pull request merged**